rcl/src/rcl/logging_rosout.c

```diff
@@ -87,48 +98,138 @@ rcl_ret_t rcl_logging_rosout_init(const rcl_allocator_t * allocator)
   char * key = NULL;
   rosout_map_entry_t entry;
-
-  // fini all the outstanding publishers
   rcutils_ret_t hashmap_ret = rcutils_hash_map_get_next_key_and_data(
     &__logger_map, NULL, &key, &entry);
   while (RCL_RET_OK == status && RCUTILS_RET_OK == hashmap_ret) {
-    // Teardown publisher
-    status = rcl_publisher_fini(&entry.publisher, entry.node);
-
-    if (RCL_RET_OK == status) {
+    if (entry.node == node) {
       status = rcl_ret_from_rcutils_ret(rcutils_hash_map_unset(&__logger_map, &key));
+      previous_key = NULL;
+    } else {
+      previous_key = key;
     }
-
     if (RCL_RET_OK == status) {
-      hashmap_ret = rcutils_hash_map_get_next_key_and_data(&__logger_map, NULL, &key, &entry);
+      hashmap_ret = rcutils_hash_map_get_next_key_and_data(
+        &__logger_map, previous_key ? &previous_key : NULL, &key, &entry);
+    }
+  }
+  return RCL_RET_OK;
+}
+
+static rcl_ret_t
+_rcl_logging_rosout_clear_logger_map_item(void * value)
+{
+  rosout_map_entry_t * entry = (rosout_map_entry_t *)value;
+  // Teardown publisher
+  rcl_ret_t status = rcl_publisher_fini(&entry->publisher, entry->node);
+  if (RCL_RET_OK == status) {
+    // delete all entries using this node
+    status = rcl_ret_from_rcutils_ret(_rcl_logging_rosout_remove_logger_map(entry->node));
+  }
+
+  return status;
+}
+
+static rcl_ret_t
+_rcl_logging_rosout_clear_sublogger_map_item(void * value)
+{
+  rosout_sublogger_entry_t * entry = (rosout_sublogger_entry_t *)value;
+  rcl_ret_t status = rcl_ret_from_rcutils_ret(
+    rcutils_hash_map_unset(&__sublogger_map, &entry->name));
+  __rosout_allocator.deallocate(entry->name, __rosout_allocator.state);
+  __rosout_allocator.deallocate(entry->count, __rosout_allocator.state);
+
+  return status;
+}
+
+static rcl_ret_t
+_rcl_logging_rosout_clear_hashmap(
+  rcutils_hash_map_t * map, rcl_ret_t (* predicate)(void *), void * entry)
+{
+  rcl_ret_t status = RCL_RET_OK;
+  char * key = NULL;
+
+  rcutils_ret_t hashmap_ret = rcutils_hash_map_get_next_key_and_data(
+    map, NULL, &key, entry);
+  while (RCUTILS_RET_OK == hashmap_ret) {
+    status = predicate(entry);
+    if (RCL_RET_OK != status) {
+      break;
     }
+
+    hashmap_ret = rcutils_hash_map_get_next_key_and_data(map, NULL, &key, entry);
   }
   if (RCUTILS_RET_HASH_MAP_NO_MORE_ENTRIES != hashmap_ret) {
     status = rcl_ret_from_rcutils_ret(hashmap_ret);
   }
 
   if (RCL_RET_OK == status) {
-    status = rcl_ret_from_rcutils_ret(rcutils_hash_map_fini(&__logger_map));
+    status = rcl_ret_from_rcutils_ret(rcutils_hash_map_fini(map));
+  }
+
+  return status;
+}
+
+rcl_ret_t rcl_logging_rosout_fini()
+{
+  if (!__is_initialized) {
+    return RCL_RET_OK;
+  }
+  rcl_ret_t status = RCL_RET_OK;
+  rosout_map_entry_t entry;
+  rosout_sublogger_entry_t sublogger_entry;
+
+  status = _rcl_logging_rosout_clear_hashmap(
+    &__logger_map, _rcl_logging_rosout_clear_logger_map_item, &entry);
+  if (RCL_RET_OK != status) {
+    goto exit;
```

Comment:
This is purely stylistic, but I think that we can just `return status` here.  That is, the extra `goto` here doesn't really buy us much.  Same goes a few other times below.